src/client/Microsoft.Identity.Client/ManagedIdentity/AbstractManagedIdentity.cs

```diff
@@ -136,6 +141,36 @@ internal static string GetMessageFromErrorResponse(HttpResponse response)
             return $"[Managed Identity] Error Code: {managedIdentityErrorResponse.Error} Error Message: {managedIdentityErrorResponse.ErrorDescription}";
         }
 
+        // Try to get the error message from the nested error response in case of cloud shell.
+        private string TryGetMessageFromNestedErrorResponse(string response)
+        {
+            var json = JsonHelper.ParseIntoJsonObject(response);
+
+            JsonHelper.TryGetValue(json, "error", out var error);
+            JsonHelper.TryGetValue(JsonHelper.ToJsonObject(error), "message", out var message);
+            JsonHelper.TryGetValue(JsonHelper.ToJsonObject(error), "code", out var errorCode);
+
+            if (message == null && errorCode == null)
+            {
+                _requestContext.Logger.Error(MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.");
+                return MsalErrorMessage.ManagedIdentityUnexpectedErrorResponse + $"Error response received from the server: {response}.";
+            }
+
+            StringBuilder errorMessage = new StringBuilder("[Managed Identity] ");
+
+            if (errorCode != null)
```

Comment:
This logic where you have an error code or message isn't unit tested.

Reply:
Refactored to cover all the scenarios. Also added all kinds of expected json responses known from the server [here[(https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg/msiCloudShellFix/tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHelpers.cs). Will add different error responses to unit test the error message parsing behavior.